File: jacline-lib-common/src/main/jacline/dictionary.js
goog.module("kohlschutter.coding.Dictionary");

/** @interface @suppress {checkTypes} */
class Dictionary {
    constructor() {
        this.dict = {};
    }

    containsKey(k) {
        return typeof this.dict[k] != "undefined";
    }

    get(k) {
        return this.dict[k];
    }

    put(k, v) {
        this.dict[k] = v;
    }

    remove(k) {
        delete this.dict[k];
    }

    static newDictionary() {
        return new Dictionary();
    }
}

exports = Dictionary;

File: jacline-lib-common/src/main/java/com/kohlschutter/jacline/lib/common/Dictionary.java
package com.kohlschutter.jacline.lib.common;

import java.util.HashMap;
import java.util.Map;

import com.kohlschutter.jacline.annotations.JsImplementationProvidedSeparately;

import jsinterop.annotations.JsType;

@JsType(isNative = true, namespace = "kohlschutter.coding", name = "Dictionary")
public interface Dictionary<V> {
  boolean containsKey(String k);

  V get(String k);

  void put(String k, V v);

  void remove(String k);

  @JsImplementationProvidedSeparately
  public static <V> Dictionary<V> newDictionary() {
    Map<String, V> map = new HashMap<String, V>();
    return new Dictionary<V>() {

      @Override
      public boolean containsKey(String k) {
        return map.containsKey(k);
      }

      @Override
      public V get(String k) {
        return map.get(k);
      }

      @Override
      public void put(String k, V v) {
        map.put(k, v);
      }

      @Override
      public void remove(String k) {
        map.remove(k);
      }
    };
  }
}
